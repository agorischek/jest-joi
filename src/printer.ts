import * as Joi from 'joi';
import * as chalk from 'chalk';
import { MatcherHintOptions, matcherHint } from 'jest-matcher-utils';
import stringifyObject = require('stringify-object');

export const buildMessage = function (
    pass: boolean,
    matcherName: string,
    received: unknown,
    schema: Joi.Schema,
    error: Joi.ValidationError,
    matcherHintOptions: MatcherHintOptions
) {
    const hint = matcherHint(matcherName, 'received', 'schema', matcherHintOptions);

    return pass
        ? () =>
            `${hint}\n\n` +
            `Schema:\n${expectedColor(printObject(schema.describe()))}\n\n` +
            `Received:\n${receivedColor(printObject(received))}`
        : () => `${hint}\n\nReceived:\n${error.annotate()}`;
};

export const printObject = (object: unknown) =>
    stringifyObject(object, {
        indent: '  ',
        singleQuotes: false,
    });

const receivedColor = chalk.red;
const expectedColor = chalk.green;