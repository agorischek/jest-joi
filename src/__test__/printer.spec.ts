import * as Joi from 'joi'
import { buildMessage } from "../printer";

const schema = Joi.object({ a: Joi.string().required() });

const input = { a: 'b' };
const options = {}
const { error } = schema.validate(input, options);
const matcherHintOptions = {
    isNot: true,
    promise: ""
}

test("Pass message snapshot", () => {
    const message = buildMessage(true, "toMatchSchema", input, schema, error, matcherHintOptions)()
    expect(message).toMatchSnapshot();
})

test("Fail message snapshot", () => {
    const message = buildMessage(false, "toMatchSchema", input, schema, error, matcherHintOptions)()
    expect(message).toMatchSnapshot();
})