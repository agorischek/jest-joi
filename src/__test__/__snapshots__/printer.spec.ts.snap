// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fail message snapshot 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22mtoMatchSchema[2m([22m[32mschema[39m[2m)[22m

Received:
{
  \\"a\\" [31m[1][0m: false
}
[31m
[1] \\"a\\" must be a string[0m"
`;

exports[`Pass message snapshot 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22mtoMatchSchema[2m([22m[32mschema[39m[2m)[22m

Schema:
[32m{[39m
[32m  type: \\"object\\",[39m
[32m  keys: {[39m
[32m    a: {[39m
[32m      type: \\"string\\",[39m
[32m      flags: {[39m
[32m        presence: \\"required\\"[39m
[32m      }[39m
[32m    }[39m
[32m  }[39m
[32m}[39m

Received:
[31m{[39m
[31m  a: false[39m
[31m}[39m"
`;
